### ISSUE TITLE
Marketplace AMI terminate / Stop Error

### ISSUE THREAD

**Issue by aaravindanarun** [jtriley/StarCluster#238]
$ starcluster terminate mycluster 

Try to terminate instance using starcluster command,
I got error message,

```
        Detaching volume vol-c63d77cf from master
        !!! ERROR - IncorrectInstanceState: Unable to detach volume 'vol-c63d77cf' with Marketplace codes from instance 'i-e18dffb6' as the instance is not in the 'stopped' state.
```

**Comment by jtriley** [comment]
@aaravindanarun Which marketplace AMI are you using?

**Comment by aaravindanarun** [comment]
Starcluster try to detach OS Volume before instance stopped state

Actually, instance root volume name is /dev/sda, but starcluster getting /dev/sda1. So problem was started, starcluster assuming OS Root Volume consider as another external volume. So try to unmount OS Volume.

In node.py file declare variable value is /dev/sda.

```
 @property
 def root_device_name(self):
```
-        self.instance.root_device_name = '/dev/sda'
       return self.instance.root_device_name

**Comment by aaravindanarun** [comment]
Please give me right solution of this problem. Time being i able to manage with previous mail thread solution

**Comment by jtriley** [comment]
@aaravindanarun If you have an instance of the marketplace AMI running can you please attach the output of:

```
$ GET http://169.254.169.254/latest/meta-data/block-device-mapping/root
```

**Comment by jtriley** [comment]
Basically if that command returns `/dev/sda` and _not_ `/dev/sda1` then I have a fix ready to go.

**Comment by aaravindanarun** [comment]
This command executed in master instance, 

$ GET http://169.254.169.254/latest/meta-data/block-device-mapping/root
/dev/sda1

**Comment by aaravindanarun** [comment]
Created and registered new AMI, In this AMI root_device_name is /dev/sda1. After update patches on node.py file.

I am getting error message -->
 starcluster start smallcluster 
StarCluster - (http://web.mit.edu/starcluster) (v. 0.9999)
Software Tools for Academics and Researchers (STAR)
Please submit bug reports to starcluster@mit.edu

> > > Using default cluster template: smallcluster
> > > Validating cluster template settings...
> > > Cluster template settings are valid
> > > Starting cluster...
> > > Launching a 1-node cluster...
> > > Creating security group @sc-smallcluster...
> > > Opening tcp port range 80-80 for CIDR 0.0.0.0/0
> > > Reservation:r-21b15c77
> > > Waiting for cluster to come up... (updating every 15s)
> > > 1/1 |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| 100%  
> > > !!! ERROR - Incompatible Cluster: smallcluster
> > > !!! ERROR - 
> > > !!! ERROR - The cluster 'smallcluster' was either created by a previous
> > > !!! ERROR - stable or development version of StarCluster or you manually
> > > !!! ERROR - created the '@sc-smallcluster' group. In any case
> > > !!! ERROR - 'smallcluster' cannot be used with this version of
> > > !!! ERROR - StarCluster (0.9999).
> > > !!! ERROR - 
> > > !!! ERROR - The cluster 'smallcluster' currently has 1 active nodes.
> > > !!! ERROR - 
> > > !!! ERROR - Please terminate the cluster using:
> > > !!! ERROR - 
> > > 
> > > ## !!! ERROR -     $ starcluster terminate smallcluster

In Log file,

2013-06-25 17:27:43,102 PID: 31971 config.py:551 - DEBUG - Loading config
2013-06-25 17:27:43,103 PID: 31971 config.py:118 - DEBUG - Loading file: /home/aravindan/.starcluster/config
2013-06-25 17:27:43,128 PID: 31971 awsutils.py:54 - DEBUG - creating self._conn w/ connection_authenticator kwargs = {'proxy_user': None, 'proxy_pass': None, 'proxy_port': None, 'proxy': None, 'is_secure': True, 'path': '/', 'region': RegionInfo:ap-southeast-1, 'port': None}
2013-06-25 17:27:43,777 PID: 31971 start.py:176 - INFO - Using default cluster template: smallcluster
2013-06-25 17:27:43,780 PID: 31971 utils.py:464 - DEBUG - nargs = 1
2013-06-25 17:27:43,780 PID: 31971 utils.py:465 - DEBUG - ndefaults = 1
2013-06-25 17:27:43,781 PID: 31971 utils.py:466 - DEBUG - nrequired = 0
2013-06-25 17:27:43,781 PID: 31971 utils.py:467 - DEBUG - args = []
2013-06-25 17:27:43,781 PID: 31971 utils.py:468 - DEBUG - kwargs = ['master_is_exec_host']
2013-06-25 17:27:43,782 PID: 31971 utils.py:469 - DEBUG - defaults = (True,)
2013-06-25 17:27:43,782 PID: 31971 cluster.py:507 - DEBUG - config_args = []
2013-06-25 17:27:43,783 PID: 31971 cluster.py:516 - DEBUG - config_kwargs = {'master_is_exec_host': 'False'}
2013-06-25 17:27:43,784 PID: 31971 cluster.py:1550 - INFO - Validating cluster template settings...
2013-06-25 17:27:44,760 PID: 31971 cluster.py:1566 - INFO - Cluster template settings are valid
2013-06-25 17:27:44,761 PID: 31971 cluster.py:1436 - INFO - Starting cluster...
2013-06-25 17:27:44,761 PID: 31971 cluster.py:964 - INFO - Launching a 1-node cluster...
2013-06-25 17:27:44,763 PID: 31971 cluster.py:991 - DEBUG - Launching master (ami: ami-a2246df0, type: t1.micro)
2013-06-25 17:27:44,878 PID: 31971 awsutils.py:165 - INFO - Creating security group @sc-smallcluster...
2013-06-25 17:27:46,018 PID: 31971 cluster.py:631 - INFO - Opening tcp port range 80-80 for CIDR 0.0.0.0/0
2013-06-25 17:27:47,390 PID: 31971 cluster.py:780 - INFO - Reservation:r-21b15c77
2013-06-25 17:27:47,393 PID: 31971 cluster.py:1244 - INFO - Waiting for cluster to come up... (updating every 15s)
2013-06-25 17:27:47,741 PID: 31971 cluster.py:671 - DEBUG - existing nodes: {}
2013-06-25 17:27:47,742 PID: 31971 cluster.py:679 - DEBUG - adding node i-c1779697 to self._nodes list
2013-06-25 17:27:48,304 PID: 31971 cli.py:279 - ERROR - Incompatible Cluster: smallcluster

The cluster 'smallcluster' was either created by a previous stable or development version of StarCluster or you manually created the '@sc-smallcluster' group. In any case 'smallcluster' cannot be used with this version of StarCluster (0.9999).

The cluster 'smallcluster' currently has 1 active nodes.

Please terminate the cluster using:

```
$ starcluster terminate smallcluster
```

**Comment by jtriley** [comment]
When you create the AMI you need to make sure that the root device name matches the block device mapping - this is what's causing the entire issue both with ebsimage command and with stop/terminate. I have another fix coming that tries to handle AMIs with broken root device info.

With that said, the IncompatibleCluster issue is really strange and I'm not sure how that happened. Looking at the code an exception should also have been logged. Can you please check your logs for a message with PID: 31971 that looks like:

```
2013-06-25 17:27:47,742 PID: 31971 cluster.py - DEBUG - load receipt exception:
(exception here)
```

If you can paste that it should give more details for debugging.

**Comment by aaravindanarun** [comment]
Thanks for reply, Thanks for providing Amazon AWS reference link.

I am tested switch root device name from /dev/sda to /dev/sda1 - without update any patch the starcluster was working stop and terminate running instance on CentOS AMI.

I assumed CentOS Marketplace AMI root device name is /dev/sda. 
An idea to switch the root device from /dev/sda to /dev/sda1 on using CentOS Marketplace AMI. It's default root device value in Amazon AWS.

Steps:
1. launch an instance from 'CentOS 6.3 Market place' AMI. (actually in my case ami-34226166 in Singapore Region)
2. Edit /boot/grub/grub.conf file,  /dev/sda to /dev/sda1.
3. Create a snapshot of root device EBS volume.
4. register an AMI image with the snapshot with the following command,
    ec2-register --region XX-XXXXXXXX-X -n 'XXXXX' -d XXXXX --root-device-name
    /dev/sda1 -b /dev/sda1=snap-xxxxxxxx -a x86_64 --kernel aki-xxxxxxxx

Launch created AMI using starcluster, then try with stop / terminate command:

It's Working for me.

**Comment by jtriley** [comment]
@aaravindanarun Glad you got this figured out. I'm not surprised that after you fixed the AMI these patches were not needed. As I mentioned this issue is entirely based on the fact that the AMI's root device name does not exist in the image's block device mapping. I'm going to go ahead and implement my fix (I accidentally closed #238 instead of #258) that tries to handle this for the case that /dev/sda1 is not in the EBS mapping and /dev/sda is given that other folks may try to extend the CentOS 6.x AMIs and run into similar issues when stopping/terminating clusters.
